src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12918,6 +12925,31 @@ private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToke
                 expected);
         }
 
+        private bool IsValidInitializerToken(ref SyntaxToken startToken, Func<bool> isInitializerMemberFn)
```

Comment:
I'm having a little difficulty figuring out the purpose of everything in this helper. It seems odd to me that we need to use a separate reset point in here. I'm also concerned that we're introducing a capture of `this` in the delegate argument. If we decide to keep this factoring it will be better to have `Func<LanguageParser, bool>` which takes `this` as an argument, and static lambdas at the call site.

I am wondering if the implementation of switch expression arm parsing would be helpful to review as we consider how to factor this.
https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs#L624